nexus/db-queries/src/db/queries/region_allocation.rs

```diff
@@ -160,6 +159,30 @@ pub fn allocation_query(
     }
     .sql("),");
 
+    // Any zpool already have this volume's existing regions?
+    builder.sql("
+  existing_zpools AS (
+    SELECT
+      dataset.pool_id
+    FROM
+      dataset INNER JOIN old_regions ON (old_regions.dataset_id = dataset.id)
+  ),")
+
+    // Filter out candidate zpools if any of a volume's old regions are already
+    // allocated to it.
+    .sql("
+  candidate_zpools_filtered AS (
+   SELECT
+     candidate_zpools.pool_id
+   FROM
+     candidate_zpools
+   WHERE (
+     NOT(
+       candidate_zpools.pool_id = ANY(SELECT existing_zpools.pool_id FROM existing_zpools)
+     )
+   )
+  ),")
```

Comment:
I think this works, but fwiw, I suspect `candidate_zpools_filtered` and `candidate_zpools` will practically be the same during region replacement, as `candidate_zpools` already filters out sleds (and will filter out disks, in https://github.com/oxidecomputer/omicron/pull/5335/)  that are not "in service + active". And I think we're probably setting that state before running region replacement anyway?

Anyway, it doesn't seem like we *want* to use `candidate_zpools` anywhere in this CTE, other than to generate `candidate_zpools_filtered`. Maybe we should add a `JOIN` + `WHERE` to `candidate_zpools` to just make it the set we want?

Reply:
agreed, done in a083bb861

Reply:
Looks great, thank you!